validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/RecentChainDataPerformanceTracker.java

```diff
@@ -0,0 +1,261 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.coordinator.performance;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.compute_epoch_at_slot;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.compute_start_slot_at_epoch;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_block_root;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_block_root_at_slot;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.IntSummaryStatistics;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.blocks.BeaconBlock;
+import tech.pegasys.teku.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.datastructures.operations.Attestation;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.logging.StatusLogger;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+public class RecentChainDataPerformanceTracker implements PerformanceTracker {
+
+  @VisibleForTesting
+  final NavigableMap<UInt64, Set<SignedBeaconBlock>> sentBlocksByEpoch = new TreeMap<>();
+
+  @VisibleForTesting
+  final NavigableMap<UInt64, Set<Attestation>> sentAttestationsByEpoch = new TreeMap<>();
+
+  @VisibleForTesting
+  static final UInt64 BLOCK_PERFORMANCE_EVALUATION_INTERVAL = UInt64.valueOf(2); // epochs
+
+  private final RecentChainData recentChainData;
+  private final StatusLogger statusLogger;
+  private Optional<UInt64> nodeStartEpoch = Optional.empty();
+
+  public RecentChainDataPerformanceTracker(
+      RecentChainData recentChainData, StatusLogger statusLogger) {
+    this.recentChainData = recentChainData;
+    this.statusLogger = statusLogger;
+  }
+
+  @Override
+  public void start(UInt64 nodeStartSlot) {
+    this.nodeStartEpoch = Optional.of(compute_epoch_at_slot(nodeStartSlot));
+  }
+
+  @Override
+  public void onSlot(UInt64 slot) {
+    if (nodeStartEpoch.isEmpty()) {
+      return;
+    }
+
+    UInt64 currentEpoch = compute_epoch_at_slot(slot);
+    if (!compute_start_slot_at_epoch(currentEpoch).equals(slot)) {
+      return;
+    }
```

Comment:
is this just an expensive modulo operation?
```
if (slot.mod(SLOTS_PER_EPOCH).isGreaterThan(ZERO)) {
return;
}
```

Reply:
Done.